VMware.CloudFoundation.CertificateManagement.psm1

```diff
@@ -25,17 +25,924 @@ if ($PSEdition -eq 'Desktop') {
+                "VC", "VE", "VG", "VI", "VN", "VU", "WF", "WS", "YE", "YT", "ZA", "ZM", "COM", "EDU", "GOV", "INT", "MIL", "NET", "ORG", "ARPA")] [String]$Country,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String] $Locality,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String] $Organization,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String] $OrganizationUnit,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String] $StateOrProvince
+    
+    )
+    
+    Try {
+        if (!(Test-Path $outputFolder)) {
+            Write-Error "Please specify a valid directory to save the CSR files." -ErrorAction Stop
+            return
+        }
+        $vcfVcenterDetails = Get-vCenterServerConnection -server $server -user $user -pass $pass -domain $domain 
+        if ($PsBoundParameters.ContainsKey("cluster")) {
+            if (Get-Cluster | Where-Object { $_.Name -eq $cluster }) {
+                $esxiHosts = Get-Cluster $cluster | Get-VMHost | Sort-Object -Property Name
+                if (!$esxiHosts) { Write-Warning "No ESXi hosts found within $cluster cluster." }
+            }
+            else {
+                Write-Error "Unable to locate Cluster $cluster in $($vcfVcenterDetails.fqdn) vCenter Server: PRE_VALIDATION_FAILED"
+                Throw "Unable to locate Cluster $cluster in $($vcfVcenterDetails.fqdn) vCenter Server: PRE_VALIDATION_FAILED"
+            }
+        }
+        else {
+            $esxiHosts = Get-VMHost -Name $esxiFqdn
+            if (!$esxiHosts) { Write-Warning "No ESXi host '$esxiFqdn' found within workload domain '$domain'." }
+        }
+
+        if ($esxiHosts) {
+            Foreach ($esxiHost in $esxiHosts) {
+                $csrPath = "$outputFolder\$($esxiHost.Name).csr"
+                $esxRequest = New-VIMachineCertificateSigningRequest -Server $vcfVcenterDetails.fqdn -VMHost $esxiHost.Name -Country "$Country" -Locality "$Locality" -Organization "$Organization" -OrganizationUnit "$OrganizationUnit" -StateOrProvince "$StateOrProvince" -CommonName $esxiHost.Name
+                $esxRequest.CertificateRequestPEM | Out-File $csrPath -Force
+                if (Test-Path $csrPath -PathType Leaf ) {
+                    Write-Host "CSR for $($esxiHost.Name) has been generated and saved to $csrPath"
+                }
+                else {
+                    Write-Error "Unable to generate CSR for $($esxiHost.Name)."
+                    Throw "Unable to generate CSR for $($esxiHost.Name)"
+                }
+            }
+        }                        
+    }
+    Catch {
+        Debug-ExceptionWriter -object $_
+    }
+    Finally {
+        Disconnect-VIServer $vcfVcenterDetails.fqdn -Confirm:$false -WarningAction SilentlyContinue
+    }
+}
+Export-ModuleMember -Function Get-EsxiCSR
+
+
+Function Get-CertManagementModeForESXi {
```

Comment:
Would this be better `Get-vCenterCertManagementMode` or `Get-vCenterCertificateManagementMode`? 

This is for vCenter Server and typically the object from which it's being requested comes first in the function name.

Reply:
Changed to Get-vCenterCertManagementMode